File: py/wsg.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
 
# Copyright 2014 Sebastien Bechet
 
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License Version 3 as
#  published by the Free Software Foundation.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
 
# [ms][Pac] WSG Sound Generator and more ....
# High Quality - version 20140714
 
# Thank you :
# Frederic Vecoven (frederic@vecoven.com) (http://www.vecoven.com/elec/pacman/pacman.html)
 
import array, wave, os
from wildcardman import MusicsAndEffects
 
# debug
#import pdb
# breakpoint: pdb.set_trace()
 
class Track(object):
    wave = 0
    frequency = 0
    voltype = 0
    volume = 0
    counter = 0
    accumulator = 0
 
    def __init__(self):
        self.counter = 0
        self.frequency = 0
        self.volume = 0
        self.voltype = 0
        self.accumulator = 0
 
    def volumeEffect(self):
        #if self.voltype == 0: # constant volume
        #    return self.volume
        if self.volume > 0:
            if self.voltype == 1: # decreasing volume
                self.volume -= 1
            elif self.voltype == 2: # -1 per 1/2 rate
                if self.counter%2:
                    self.volume -= 1
        self.counter += 1
        return self.volume
 
    def update(self):
        if self.volume == 0 or self.frequency == 0:
            self.accumulator = 0
 
    def isFinished(self):
        return False
 
class Effect(Track):
    def __init__(self, source):
        Track.__init__(self)
        # Data Effect (ro)
        self.duration = source[3]
        self.reverse = source[4]
        self.freqincrepeat = source[5]
        self.repeat = source[6]
        self.volinc = source[9]
        # Helper (rw)
        self.freqstart = source[1]
        self.freqinc = source[2]
        self.play = True
        self.pingpong = True
        self.curduration = self.duration
        # Track (rw)
        self.wave = source[0]
        self.frequency = self.freqstart
        self.voltype = source[7]
        self.volume = source[8]
 
    def update(self):
        Track.update(self)
        if not self.play:
            return False
 
        self.curduration -= 1;
        if self.curduration == 0:
            # duration is now 0, check repeat
            self.repeat -= 1
            if self.repeat <= 0:
                self.frequency = 0
                self.volume = 0
                self.play = False
                return False
            # duration is 0. Re-init it.
            self.curduration = self.duration
            # Reverse Frequency increment
            if self.reverse:
                self.freqinc = - self.freqinc
                self.pingpong = not self.pingpong
            if (not self.reverse) or self.pingpong:
                self.freqstart += self.freqincrepeat
                self.frequency = self.freqstart
                self.volume += self.volinc
 
        self.frequency += self.freqinc
        self.volumeEffect()
        return True
 
    def isFinished(self):
        return not self.play
 
class Music(Track):
    notes = {}
 
    def __init__(self, source):
        Track.__init__(self)
        # Data Effect (ro)
        self.data = source
        # Helper (rw)
        self.play = True
        self.duration = 1
        self.seek = 0
        # Track (rw)
        self.wave = 0
        self.frequency = 0
        self.voltype = 0
        self.volume = 0
        self.initNote()
 
    def initNote(self):
        # No Music
        k2 = '---'
        v2 = 0
        self.notes[k2] = v2
        for k, v in MusicsAndEffects.octave.items():
            k2 = k + '0'
            v2 = int(v / 8)
            self.notes[k2] = v2
            k2 = k + '1'
            v2 = int(v / 4)
            self.notes[k2] = v2
            k2 = k + '2'
            v2 = int(v / 2)
            self.notes[k2] = v2
            k2 = k + '3'
            v2 = int(v)
            self.notes[k2] = v2
            for i in (1, 2, 3, 4, 5, 6):
                k2 = k + str(i+3)
                v2 = int(v * pow(2,i))
                self.notes[k2] = v2
 
    def update(self):
        Track.update(self)
        if not self.play:
            return False
 
        #wave, note, duration (1/60 s), volume, volEffect
        self.duration -= 1
        if self.duration == 0:
            if self.seek >= len(self.data):
                self.play = False
                return False
            self.wave = self.data[self.seek][0]
            self.note = self.data[self.seek][1]
            self.duration = self.data[self.seek][2]
            self.volume = self.data[self.seek][3]
            self.voltype = self.data[self.seek][4]
            self.seek += 1
 
            self.frequency = self.notes[self.note]
 
        self.volumeEffect()
        return True
 
    def isFinished(self):
        return not self.play
 
#----------------------------------------------------------------------
class Wsg(object):
    def __init__(self, bank, outputFrequency):
        self.bank = bank
        self.count = 0
        self.outputFrequency = outputFrequency
        self.vblank = int(outputFrequency / 60)
        self.wavetable = MusicsAndEffects.wavetable
 
    def update(self,length):
        sequence = length * [0]
        better_size_is = 0   # real sequence length
        for track in self.bank:
            incr = track.frequency / self.outputFrequency
            wtab = self.wavetable[track.wave]
            for i in range(length):
                if track.isFinished():
                    break
                if self.count%self.vblank == 0:
                    track.update()
                    incr = track.frequency / self.outputFrequency
                if (track.volume != 0) and (track.frequency != 0):
                    seek = int(track.accumulator) % 32
                    sequence[i] += track.volume * wtab[abs(seek)]
                    track.accumulator += incr
                self.count += 1
            if i > better_size_is:
                better_size_is = i + 1
        return sequence[:better_size_is]
 
    def isFinished(self):
        finished = True;
        for track in self.bank:
            finished = track.isFinished() & finished
        return finished
#----------------------------------------------------------------------
class SoundWave(Wsg):
    wave_file = None
 
    def __init__(self, bank, outputFrequency, name):
        Wsg.__init__(self, bank, outputFrequency)
        # SHRT_MAX / (max(sample)*max(volume)*nbtrack)
        self.normalize = int((2**15) / (16*16*len(bank)))
        self.wave_file = wave.open(name + ".wav", "w")
        nchannels = 1
        sampwidth = 2
        framerate = outputFrequency
        nframes = 0 # this is a library problem, not my own
        comptype = "NONE"
        compname = "not compressed"
        self.wave_file.setparams((nchannels, sampwidth, framerate,
            nframes, comptype, compname))
 
    def __del__(self):
        self.wave_file.close()
 
    def update(self):
        seq = Wsg.update(self,4096)
        seq16 = [s * self.normalize for s in seq] # normalize
        seqb = array.array('h', seq16).tostring() # then in bytes()
        self.wave_file.writeframes(seqb)
 
def genEffect(destFreq):
    for name, data in MusicsAndEffects.effects.items():
        print("===" + name + "===")
        dname = 'waves/' + name
        e = []
        e.append(Effect(data))
        s = SoundWave(e, destFreq, dname)
        while not s.isFinished():
            seq = s.update()
 
def genMusic(destFreq):
    for name, tracks in MusicsAndEffects.musics.items():
        print("===" + name + "===")
        dname = 'waves/' + name
        m = []
        for track in tracks:
            m.append(Music(track))
        s = SoundWave(m, destFreq, dname)
        while not s.isFinished():
            seq = s.update()
 
if __name__ == '__main__':
    try:
        os.mkdir('waves')
    except OSError as e:
        print('wave/ : ', e.args)
    destFreq = 48000
    genEffect(destFreq)
    genMusic(destFreq)

File: js/wsg.js
// -*- coding: utf-8 -*-
// Copyright 2014 Sebastien Bechet

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License Version 3 as
//  published by the Free Software Foundation.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.

// [ms][Pac] WSG Sound Generator and more ....
// Javascript version 20140630

// Thank you :
// Frederic Vecoven (frederic@vecoven.com) (http://www.vecoven.com/elec/pacman/pacman.html)

///////////////////////////////////////////////////////////////////////
// Track

var Track = function(name) {
    this.name = name;
    this.wave = 0;
    this.frequency = 0;
    this.voltype = 0;
    this.volume = 0;
    this.counter = 0;
    this.accumulator = 0;
    this.play = false;
};

Track.prototype = {
    volumeEffect: function() {
        if (this.volume > 0) {
            if (this.voltype == 1) { // decreasing volume
                this.volume -= 1;
            } else {
                if (this.voltype == 2) { // -1 per 1/2 rate
                    if (this.counter % 2) {
                        this.volume -= 1;
                    }
                }
            }
        }
        this.counter += 1;
        return this.volume;
    },
    update: function() {
        if (this.volume == 0 || this.frequency == 0) {
            this.accumulator = 0;
        }
    },
    isFinished: function() {
        return ! this.play;
    }
};

///////////////////////////////////////////////////////////////////////
// Effect

var Effect = function(name, source) {
    Track.call(this, name);
    // Data Effect (ro)
    this.duration = source[3];
    this.reverse = source[4];
    this.freqincrepeat = source[5];
    this.repeat = source[6];
    this.volinc = source[9];
    // Helper (rw)
    this.freqstart = source[1];
    this.freqinc = source[2];
    this.play = true;
    this.pingpong = true;
    this.curduration = this.duration;
    // Track (rw)
    this.wave = source[0];
    this.frequency = this.freqstart;
    this.voltype = source[7];
    this.volume = source[8];
};

Effect.prototype = newChildObject(Track.prototype, {
    update: function() {
        Track.prototype.update.call(this);
        if (! this.play) {
            return false;
        }
        this.curduration -= 1;
        if (this.curduration == 0) {
            // duration is now 0, check repeat
            this.repeat -= 1;
            if (this.repeat <= 0) {
                this.frequency = 0;
                this.volume = 0;
                this.play = false;
                return false;
            }
            // duration is 0. Re-init it.
            this.curduration = this.duration;
            // Reverse Frequency increment
            if (this.reverse) {
                this.freqinc = - this.freqinc;
                this.pingpong = !this.pingpong;
            }
            if ( (!this.reverse) || this.pingpong) {
                this.freqstart += this.freqincrepeat;
                this.frequency = this.freqstart;
                this.volume += this.volinc;
            }
        }
        this.frequency += this.freqinc;
        Track.prototype.volumeEffect.call(this);
        return true;
    },
    isFinished: function() {
        return ! this.play;
    }
});

///////////////////////////////////////////////////////////////////////
// Music

var Music = function(name, source) {
    Track.call(this, name);
    // Data Effect (ro)
    this.data = source;
    // Helper (rw)
    this.play = true;
    this.duration = 1;
    this.seek = 0;
    // Track (rw)
    this.wave = 0;
    this.frequency = 0;
    this.voltype = 0;
    this.volume = 0;
    // initNote :
    var octave = [];
    // octave number 4 in real world
    octave["C-"] = Math.round(4*440/Math.pow(2,9/12));
    octave["C#"] = Math.round(4*440/Math.pow(2,8/12));
    octave["D-"] = Math.round(4*440/Math.pow(2,7/12)); 
    octave["D#"] = Math.round(4*440/Math.pow(2,6/12)); 
    octave["E-"] = Math.round(4*440/Math.pow(2,5/12));
    octave["F-"] = Math.round(4*440/Math.pow(2,4/12));
    octave["F#"] = Math.round(4*440/Math.pow(2,3/12));
    octave["G-"] = Math.round(4*440/Math.pow(2,2/12));
    octave["G#"] = Math.round(4*440/Math.pow(2,1/12));
    octave["A-"] = Math.round(4*440);
    octave["A#"] = Math.round(4*440*Math.pow(2,1/12));
    octave["B-"] = Math.round(4*440*Math.pow(2,2/12));
    this.notes = {};
    var k2, v2, v;
    // No Music
    k2 = '---';
    v2 = 0;
    this.notes[k2] = v2;
    for (var k in octave) {
        v = octave[k]
        k2 = k + '0';
        v2 = v / 8;
        this.notes[k2] = v2;
        k2 = k + '1';
        v2 = v / 4;
        this.notes[k2] = v2;
        k2 = k + '2';
        v2 = v / 2;
        this.notes[k2] = v2;
        k2 = k + '3';
        v2 = v;
        this.notes[k2] = v2;
        for (var i=1; i<6; i++) {
            k2 = k + (i+3).toString();
            v2 = v * Math.pow(2,i);
            this.notes[k2] = v2;
        }
    }
};

Music.prototype = newChildObject(Track.prototype, {
    update: function() {
        Track.prototype.update.call(this);
        if (! this.play) {
            return false;
        }
        //wave, note, duration (1/60 s), volume, volEffect
        this.duration -= 1;
        if (this.duration == 0) {
            if (this.seek >= this.data.length) {
                this.frequency = 0;
                this.volume = 0;
                this.play = false;
                return false;
            }
            this.wave = this.data[this.seek][0];
            this.note = this.data[this.seek][1];
            this.duration = this.data[this.seek][2];
            this.volume = this.data[this.seek][3];
            this.voltype = this.data[this.seek][4];
            this.seek += 1;
            this.frequency = this.notes[this.note];
        }
        Track.prototype.volumeEffect.call(this);
        return true
    },
    isFinished: function() {
        return ! this.play;
    }
});

///////////////////////////////////////////////////////////////////////
// Wsg

function Wsg(outputFrequency) {
    var wavetable = MusicsAndEffects.wavetable;
    var bank = [];
    var seqtrack = [];
    var count = 0;
    var outputFrequency = outputFrequency;
    var vblank = Math.round(outputFrequency / 60);

    this.update = function(e) {
        var sequence = e.outputBuffer.getChannelData(0);
        var len = sequence.length;

        // cleanup buffer (chromium)
        for (var i=0; i<len; i++) {
            sequence[i] = 0;
        }
        
        // remove old tracks
        for (var i=0; i<bank.length; i++) {
            var track = bank[i];
            if (track.isFinished()) {
                bank.splice(bank.indexOf(track),1);
            }
        }
        
        // add sequence
        if (bank.length == 0) {
            if (seqtrack.length != 0) {
                bank.push(seqtrack.shift());
            }
        }

        // fill samples
        for (var index=0; index<bank.length; index++) {
            var track = bank[index];
            var incr = track.frequency / outputFrequency;
            var wtab = wavetable[track.wave];
            for (var i=0; i<len; i++) {
                if (count % vblank == 0) {
                    if (track.isFinished()) {
                        break;
                    } else {
                        track.update();
                        incr = track.frequency / outputFrequency;
                    }
                }                
                if ( (track.volume != 0) || (track.frequency != 0) ) {
                    var seek = Math.floor(track.accumulator) % 32;
                    sequence[i] += track.volume * wtab[Math.abs(seek)];
                    track.accumulator += incr;
                }
                count += 1;
            }
        }
        
        // normalize [-1 ... 1] (volume and tracks number)
        if (bank.length != 0) {
            for (var i=0; i<len; i++) {
                sequence[i] /= (15*3); // divide by three voice, else sirens are too high
            }
        }
    }
    this.addTrack = function(track) {
        bank.push(track);
        return true;
    }
    
    this.addSeqTrack = function(track) {
        seqtrack.push(track);
    }

    this.resetSeqTrack = function(track) {
        seqtrack = [];
    }
    
    this.isFinished = function() {
        var finished = true;
        for (var index=0; index<bank.length; index++) {
            var track = bank[index];
            finished = track.isFinished() && finished;
        }
        return finished;
    }

    this.isPlaying = function(name) {
        // Test if already playing
        for (var i=0; i<bank.length; i++) {
            if (bank[i].name == name) {
                return true;
            }
        }
        return false;
    }

    // Correct samples for matching [-1 .. 1]
    for (var i=0; i<wavetable.length; i++) {
        for (var j=0; j<wavetable[i].length; j++) {
            wavetable[i][j] = ((wavetable[i][j]+8)/7.5) - 1;
        }
    }
    return this;
}


///////////////////////////////////////////////////////////////////////
// WebAudio API

getWsgOnWebAudio = function() {
    try {
        var audio_ctx = new (window.AudioContext || window.webkitAudioContext)();
    } catch (e) {
        alert("Sorry, html5 audio not supported.");
    }
    //~ var gain = audio_ctx.createGain();
    //~ gain.gain.value = 0.5;

    var wsg = new Wsg(audio_ctx.sampleRate);

    var processor = audio_ctx.createScriptProcessor(4096, 1, 1);
    processor.onaudioprocess = wsg.update;
    processor.connect(audio_ctx.destination);
    //~ processor.connect(gain);
    //~ gain.connect(audio_ctx.destination);
    return wsg;
}

addMusic = function(name) {
    if (wsg.isPlaying(name)) {
        return;
    }
    var tracks=MusicsAndEffects.musics[name];
    for (var i=0; i<tracks.length; i++) {
        var track = tracks[i];
        var t = new Music(name, track);
        wsg.addTrack(t);
    }
}

addEffect = function(name) {
    if (wsg.isPlaying(name)) {
        return;
    }
    var effect=MusicsAndEffects.effects[name];
    var t = new Effect(name, effect);
    wsg.addTrack(t);
}

addSeqEffect = function(name) {
    var effect=MusicsAndEffects.effects[name];
    var t = new Effect(name, effect);
    wsg.addSeqTrack(t);
}

addNewSeqEffect = function(name) {
    wsg.resetSeqTrack();
    addSeqEffect(name);
}
